### PR TITLE
refactor(runtime): convert more dist#ft functions to lua

### PR DESCRIPTION
**Pull request by smjonas**
Follow-up to #18247 (working on #18241) .

### CODE REVIEW

**Comment by smjonas** [comment]
Converting the functions is going pretty smooth, however I think we should establish some rules to follow. Please let me know how you think these cases should be handled. This should probably be documented somewhere as well?
1.  Vim often uses `\<` and `\>` for start / end of a word. Suggestion: if this occurs at the end of a pattern, simply omit it. If in the middle, use a `.` instead.
2. Usage of `=~`. Suggestion: assume `ignorecase` (i.e. convert the line to lowercase first and check afterwards).
3. Usage of `\k`. Suggestion: simply use `.`.

What do you think?

I don't think any of these changes will lead to test failures as the tests are not that detailed.

**Comment by smjonas** [comment]
@gpanders Should I stop adding new functions here and leave the remaining 9 for round three? :smile:

**Comment by gpanders** [comment]
Yes that sounds good, it looks like there is already enough here to keep me busy reviewing 😄

**Comment by smjonas** [comment]
That's definitely true😅I will let you know once all tests pass and everything looks good to me

**Comment by clason** [comment]
> 1. Vim often uses `\<` and `\>` for start / end of a word. Suggestion: if this occurs at the end of a pattern, simply omit it. If in the middle, use a `.` instead.

I'll defer to @gpanders on this; I'm barely regex-literate.

> 2. Usage of `=~`. Suggestion: assume `ignorecase` (i.e. convert the line to lowercase first and check afterwards).

That sounds good to me.

> 3. Usage of `\k`. Suggestion: simply use `.`.

See 1.; however this is decided, I think it would be good to document this in a comment block at the top of the file. (Also previous decisions like `find` over `match` and `sub`.) This will be very valuable when new `dist#ft` functions have to be ported in the future.

> I don't think any of these changes will lead to test failures as the tests are not that detailed.

And if they do, we will know :)

> Should I stop adding new functions here and leave the remaining 9 for round three? 😄

Don't forget `scripts.vim`, which is the boss battle :P

**Comment by smjonas** [comment]
@gpanders I noticed there is no code in `filetype.lua` referencing 4 of these patterns (except one for `upstreamlog`). Which seems to be the reason why some of the tests fail. Does anything need to be added there?
```vim
" Innovation Data Processing
au BufRead,BufNewFile upstream.dat\c,upstream.*.dat\c,*.upstream.dat\c	setf upstreamdat
au BufRead,BufNewFile fdrupstream.log,upstream.log\c,upstream.*.log\c,*.upstream.log\c,UPSTREAM-*.log\c	setf upstreamlog
au BufRead,BufNewFile upstreaminstall.log\c,upstreaminstall.*.log\c,*.upstreaminstall.log\c setf upstreaminstalllog
au BufRead,BufNewFile usserver.log\c,usserver.*.log\c,*.usserver.log\c	setf usserverlog
au BufRead,BufNewFile usw2kagt.log\c,usw2kagt.*.log\c,*.usw2kagt.log\c	setf usw2kagtlog
```

**Comment by clason** [comment]
That's what https://github.com/neovim/neovim/pull/18219 is for...

**Comment by smjonas** [comment]
Oh, makes sense, so would it be better if I refrained from calling the new `dat` function for now until that PR is merged?

**Comment by clason** [comment]
Or you could add the patterns, whichever is easier.

**Comment by smjonas** [comment]
I am not really sure how I'm supposed to resolve the merge conflicts :smile: Is this even necessary or can this be done once this PR is ready to be merged? Rebasing on master  sounds simple but I'm not that experienced with Git. Please let me know :)

**Comment by clason** [comment]
I would do the following: 

1. download the stylua config directly to the `neovim/` directory from https://raw.githubusercontent.com/neovim/neovim/master/.stylua.toml
2. run (from the repo root) `stylua runtime/lua/vim/filetype.lua runtime/lua/vim/filetype/detect.lua`
3. commit the result and squash your commits
4. from the [Github page for your fork](https://github.com/smjonas/neovim/tree/master), select "Fetch upstream" (in the top bar "This branch is X commits behind...")
5. from the repo root, do `git checkout master; git pull; git checkout filetype-detect; git rebase master` (you might have to remove the `.stylua.toml` first)
6. `git push -f`

And, yes, you can do that whenever you are ready; it doesn't have to be now.

**Comment by smjonas** [comment]
With almost all tests passing (except the ones for `gomod` files, maybe you can figure out what's wrong), this is ready for a review ^^ @gpanders

**Comment by gpanders** [comment]
> 1. Vim often uses `\<` and `\>` for start / end of a word. Suggestion: if this occurs at the end of a pattern, simply omit it. If in the middle, use a `.` instead.

I've been thinking about this, and I think that we should be prioritizing correctness over performance. There's a chance that the patterns in filetype.vim that are using `\<` and `\>` don't actually *need* to be using them... but that would be very hard to verify, so I lean toward "trust the people who created the patterns in the first place, as they know the requirements of their filetypes better than I do". So I've come around to just using `vim.regex()` for cases like this. It is slower than plain Lua pattern matching, but we're not detecting millions of filetypes in a hot loop so it's ok to lose a little bit of perf if it means we're more correct.

> 2. Usage of `=~`. Suggestion: assume `ignorecase` (i.e. convert the line to lowercase first and check afterwards).

This seems like a safe assumption. Users can file an issue if this causes missed detections.

> 3. Usage of `\k`. Suggestion: simply use `.`.

See my response to point 1. I think we should just use vim.regex() for these after all.

**Comment by gpanders** [comment]
> This seems like a safe assumption. Users can file an issue if this causes missed detections.

Actually I want to amend this. I think for cases of `=~` we should use the following heuristic:

1. If the string being compared against has *any* upper case letters, treat the comparison as case *sensitive*
2. Otherwise, treat it as case-insensitive.

Basically we are applying `'smartcase'`. If the author of the code explicitly used upper case characters then it's likely that case does matter (they *should* have used `=~#` but that's an easy mistake to make in Vim script).

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:591]:
```suggestion
      return vim.regex([[\<module\s\+\w\+\s*\.\s*\(%\|$\)]]):match_str(line) ~= nil
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:602]: `MODULE` is uppercase in filetype.vim so let's assume it's case sensitive here. And since it uses `\<` let's just copy the pattern directly:

```suggestion
  elseif vim.regex([[\%(\<MODULE\s\+\w\+\s*;\|^\s*(\*\)]]):match_str(nextnonblank(bufnr, 1)) then
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:606]:
```suggestion
  elseif vim.regex([[\<go\.mod$]]):match_str(path) then
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:620]:
```suggestion
      return true
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:637]:
```suggestion
  if first_line:find('^#') and first_line:lower():find('perl') then
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:642]:
```suggestion
  local re = vim.regex([[^use\s\s*\k]])
  for _, line in ipairs(getlines(bufnr, 1, 30)) do
    if re:match_str(line) then
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:681]: ✅

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:703]:
```suggestion
      return M.asm(bufnr)
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:664]:
```suggestion
local pascal_keywords = vim.regex([[^\s*\%(program\|unit\|library\|uses\|begin\|procedure\|function\|const\|type\|var\)\>]])
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:734]:
```suggestion
    if findany(line, pascal_comments) or pascal_keywords:match_str(line:lower()) then
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:751]:
```suggestion
  if vim.regex([[\<rebol\>]]):match_str(table.concat(lines):lower()) then
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:789]: Not needed, the `path` variable is already expanded to the full path
```suggestion
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:796]:
```suggestion
      '/usr/lib/udev/.*%.rules$',
      '/usr/lib/udev/rules%.d/.*%.rules$',
      '/lib/udev/.*%.rules$',
      '/lib/udev/rules%.d/.*%.rules$',
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:811]:
```suggestion
    local dir = vim.fn.expand(path, ':h')
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:817]: Probably, but it still works since they use `setf` so a subsequent `setf` will have no effect. Returning here is the right call though.

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:876]:
```suggestion
  local re = vim.regex([[\v^\s*%(\&\w+|%(global\s+)?def%(fct)?>)]])
  if re:match_str(line) then
    vim.bo[bufnr].filetype = 'krl'
  end
```

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:879]: ✅

**Comment by gpanders** [comment]
@smjonas I also just remembered that `vim.regex()` is case insensitive by default, so there's no need to `:lower()` strings before passing them to `match_str()`.

In the cases where we need case-sensitivity, we'll need to specify `\C` at the beginning of the regex.

File: runtime/lua/vim/filetype/detect.lua
Comment (smjonas) [reply runtime/lua/vim/filetype/detect.lua:280]: Is this equivalent to `:%S-:%S-:`? If so, I think it would be clearer that way.

File: runtime/lua/vim/filetype/detect.lua
Comment (smjonas) [reply runtime/lua/vim/filetype/detect.lua:295]: Yes

File: runtime/lua/vim/filetype/detect.lua
Comment (smjonas) [reply runtime/lua/vim/filetype/detect.lua:239]: I added a missing `return` here

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [reply runtime/lua/vim/filetype/detect.lua:280]: It is! I actually did not know that Lua had that 😄

**Comment by gpanders** [comment]
@smjonas I had an idea: to avoid repeatedly re-compiling the same regex patterns over and over, we could create a cache so that each regex is only compiled once and then re-used if necessary. Something like this:

```lua
local matchre = (function()
  local cache = {}
  return function(line, regex)
    if not cache[regex] then
      cache[regex] = vim.regex(regex)
    end

    return cache[regex]:match_str(line)
  end
end)()
```

And then instances of e.g. `vim.regex([[\<some pattern\>]]):match_str(line)` could be replaced with `matchre(line, [[\<some pattern\>]])`.

`matchre` is just a name I came up with, we can use something else if you have a better idea.

File: runtime/lua/vim/filetype/detect.lua
Comment (gpanders) [inline runtime/lua/vim/filetype/detect.lua:148]: Error here when `nextnonblank` returns `nil`.


Suggestion:

```lua
local line = nextnonblank(bufnr, 1)
if line then
  return matchre(line:lower(), [[\v^\s*%(\%{3}|module\s+\k+\s*%(\(|$))]])
end
```

**Comment by gpanders** [comment]
I think the reason the tests are passing is that `vim.regex()` (in contradiction with its documentation in `:h vim.regex()` is *not* case insensitive by default. I wonder if the docs meant to say that it is case *sensitive* by default "regardless of 'ignorecase'".

In any case, I'm sorry that I misled you, I only just now discovered this fact while looking into why the tests are failing.

**Comment by smjonas** [comment]
No worries, thanks for letting me know! Let's see if the tests pass now :)

**Comment by clason** [comment]
Time for a squash, indeed!

File: runtime/lua/vim/filetype/detect.lua
Comment (clason) [inline runtime/lua/vim/filetype/detect.lua:15]:
```suggestion
--  * Sort the function alphabetically and omit 'ft' or 'check' from the new function name.
--  * Use ':find' instead of ':match' / ':sub' if possible.
--  * When '=~' is used to match a pattern, there are two possibilities:
--     - If the pattern only contains lowercase characters, treat the comparison as case-insensitive.
--     - Otherwise, treat it as case-sensitive. 
--     (Basically, we apply 'smartcase': if upper case characters are used in the original pattern, then 
--     it's likely that case does matter).
--  * When '\k', '\<' or '\>' is used in a pattern, use the 'matchregex' function.
--     Note that vim.regex is case-sensitive by default, so add the '\c' flag if only lowercase letters 
--     are present in the pattern:
--     Example:
--     `if line =~ '^\s*unwind_protect\>'` => `if matchregex(line, [[\c^\s*unwind_protect\>]])
```

**Review by gpanders — APPROVED** [review]

File: runtime/lua/vim/filetype/detect.lua
Comment (smjonas) [reply runtime/lua/vim/filetype/detect.lua]: Done

**Pull request merged**
